ops/pebble.py

```diff
@@ -1464,45 +1465,28 @@ def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.Bin
         if not boundary:
             raise ProtocolError('invalid boundary {!r}'.format(boundary))
 
-        # We have to manually write the Content-Type with boundary, because
-        # email.parser expects the entire multipart message with headers.
-        parser = email.parser.BytesFeedParser()
-        parser.feed(b'Content-Type: multipart/form-data; boundary='
-                    + boundary.encode('utf-8') + b'\r\n\r\n')
-
-        # Then read the rest of the response and feed it to the parser.
+        parser = MultipartLargeFileParser(boundary)
         while True:
             chunk = response.read(8192)
```

Comment:
It seems a bit of a shame that we have to buffer the entire content of the response into a local temporary file in order to read it. (you end up reading from disk, to write to a temp location on disk, to then read and put it somewhere else on disk).
It feels like 90% of the work has been done. All we would need to do is feed() until we got enough of the header information to know that we were reading the right content, and then return an object whose read() method returns either from the internal buffer, or calls feed() and then returns data.
That avoids tempfile and it avoids extra writes, and the in-memory buffer can stay reasonably bounded.

Reply:
FWIW, I could be wrong on this. the json payload 'response' happens after the content if I remember correctly. Which may mean we have to buffer all of the content before we can understand what file we are processing. I think I'd still rather expose that a bit to the user, that they may have to buffer into a tempfile in their target directory and rename it into the final filename, rather than buffering potentially on another disk and having to copy it yet again.

I certainly wouldn't say that this isn't an improvement over what we have, but we are so tantalizingly close, I'd like to see if it is possible before we commit to this.

Reply:
Good points from John here.

Reply:
It is a bit of a shame.  However, as long as we preserve the API and prevent callers from seeing the file on-disk, we can preserve our ability to revisit this and optimize later as necessary.  As long as we can preserve future options, I'm okay with the single-path-putting-everything-on-disk (for now).